server/src/main/java/org/elasticsearch/common/settings/ConsistentSecureSettingsValidatorService.java

```diff
@@ -0,0 +1,257 @@
+import java.util.Set;
+
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+public class ConsistentSecureSettingsValidatorService implements ClusterStateApplier, LocalNodeMasterListener {
+    private static final Logger logger = LogManager.getLogger(ConsistentSecureSettingsValidatorService.class);
+
+    private final ClusterService clusterService;
+    private final Map<String, char[]> localHashesOfConsistentSettings = new HashMap<>();
+    private volatile Map<String, String> publicHashesOfConsistentSettings = new HashMap<>();
+    private volatile boolean allSecureSettingsConsistent = true;
+
+    public ConsistentSecureSettingsValidatorService(Settings settings, ClusterService clusterService,
+                                                    Set<SecureSetting<?>> consistentSecureSettings) {
+        this.clusterService = clusterService;
+        // eagerly compute hashes for the secure settings
+        computeHashesOfLocalSecureSettingValues(settings, consistentSecureSettings);
+        // this cross checks the published hashes against the local ones
+        clusterService.addStateApplier(this);
+        // when this node is the master it publishes hashes
+        clusterService.addLocalNodeMasterListener(this);
+    }
+
+    private void computeHashesOfLocalSecureSettingValues(Settings settings, Set<SecureSetting<?>> consistentSecureSettings) {
+        for (SecureSetting<?> secureSetting : consistentSecureSettings) {
+            // compute unsalted hash to store in memory on the local node
+            final char[] localHash = computeSecureSettingLocalHash(settings, secureSetting);
+            localHashesOfConsistentSettings.put(secureSetting.getKey(), localHash);
+            // salted hash (of the unsalted hash) suitable for cluster-wide publication
+            final String publicHash = computePublicHash(localHash, UUIDs.randomBase64UUID());
+            publicHashesOfConsistentSettings.put(secureSetting.getKey(), publicHash);
+        }
+    }
+
+    public boolean allSecureSettingsConsistent() {
+        return allSecureSettingsConsistent;
+    }
+
+    /**
+     * Compute UNSALTED hash of a secure setting value for subsequent comparison with the master's reference. This values IS NOT to be
+     * published in the cluster state, but rather compared (and stored) locally, to the SALTED hash value published in the cluster state by
+     * the master node.
+     */
+    private static char[] computeSecureSettingLocalHash(Settings settings, SecureSetting<?> secureSetting) {
+        final Object verbatimValue = secureSetting.get(settings);
+        final MessageDigest digest;
+        try {
+            digest = MessageDigest.getInstance("SHA-512");
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("\"SHA-512\" algorithm is required for consistent secure settings", e);
+        }
+        if (verbatimValue instanceof InputStream) {
```

Comment:
++ I would suggest that `SecureString` extend `InputStream`. What do you say?

Reply:
Do we really need a stream? If you feel a stream is needed, then I'd rather have a method that returns a InputStream rather than making SecureString extend InputStream

Reply:
Instead of having to support both SecureSetting types here, can this be offloaded to them? For example, having a hash method on SecureSetting that will return the hash of the value. I don't think we should be calling a generic `get` on the settings for secure settings (in fact I think this should error).

Reply:
Unfortunately the interface of `SecureSetting` is generic.
@jaymode are you suggesting to enhance this interface with a method returning `byte[]` ?
@rjernst are you suggesting to enhance this interface with a method returning the hash of the value?

Would extending the interface of `SecureSettings` (notice the trailing _s_) to return the table of unsalted hashed (Map<String, String>) be a solution you both would support? I considered this solution at first, but then considered that getting the secure value through the settings infra, using the `get` method, would be more idiomatic. @rjernst It would help me to understand why you think this should error.

Reply:
> are you suggesting to enhance this interface with a method returning the hash of the value?

Yes, that is what I am suggesting. The interface of SecureSetting is generic with the intention of never having a SecureSetting; it is only there for common code and simplicity. A user within the codebase would always have a concrete T, that is returned by either secureString or secureFile. 

In regards to byte[], I do not think we should expose that. The intention in this design was to never allow an object that could be forgotten to be cleared in memory that contains secret material. For a file setting, InputStream is the correct access pattern (read over the data to find what is needed). Having a getValueHash method (or some better name) on the SecureSetting will allow the implementation to choose the best way to read the value, and ensure it is cleaned up from memory.